src/core/quantum_gate.jl

```diff
@@ -75,6 +75,106 @@ get_connected_qubits(gate::AbstractGate)=[gate.target]
 
 get_gate_parameters(gate::AbstractGate)=Dict()
 
+"""
+    move_gate(gate::AbstractGate,
+        qubit_mapping::AbstractDict{<:Integer,<:Integer})::AbstractGate
+
+Returns a copy of `gate` where the qubits on which the `gate` acts have been updated based
+on `qubit_mapping`.
+
+The dictionary `qubit_mapping` contains key-value pairs describing how to update the target
+qubits. The key indicates which target qubit to change while the associated value specifies
+the new qubit.
+
+# Examples
+```jldoctest
+julia> gate = sigma_x(1)
+Gate Object: Snowflake.SigmaX
+Connected_qubits	: [1]
+Operator:
+(2,2)-element Snowflake.AntiDiagonalOperator:
+Underlying data type: ComplexF64:
+    .    1.0 + 0.0im
+    1.0 + 0.0im    .
+
+
+julia> move_gate(gate, Dict(1=>2))
+Gate Object: Snowflake.SigmaX
+Connected_qubits	: [2]
+Operator:
+(2,2)-element Snowflake.AntiDiagonalOperator:
+Underlying data type: ComplexF64:
+    .    1.0 + 0.0im
+    1.0 + 0.0im    .
+
+
+```
+"""
+function move_gate(gate::AbstractGate,
+    qubit_mapping::AbstractDict{T,T})::AbstractGate where T<:Integer
+
+    old_connected_qubits = get_connected_qubits(gate)
+    new_connected_qubits = Int[]
+    found_move = false
+    for (i_qubit, old_qubit) in enumerate(old_connected_qubits)
+        if haskey(qubit_mapping, old_qubit)
+            if !found_move
+                new_connected_qubits = deepcopy(old_connected_qubits)
+                found_move = true
+            end
+            new_connected_qubits[i_qubit] = qubit_mapping[old_qubit]
+        end
+    end
+    if found_move
+        return move_gate(gate, new_connected_qubits)
+    else
+        return gate
+    end
+end
+
+"""
+    move_gate(gate::AbstractGate,
+        new_connected_qubits::AbstractVector{<:Integer})::AbstractGate
+
+Returns a copy of `gate` where the qubits on which the `gate` acts are
+`new_connected_qubits`.
+
+!!! tip "This function should not be called directly."
+    Because this function depends on the order of the `new_connected_qubits`, it is safer
+    to call [`move_gate(gate::AbstractGate, qubit_mapping::AbstractDict{T,T}) where T<:Integer`](@ref).
+
+This function must be overloaded for gates which are constructed with more than 1 argument.
+
+# Examples
+```jldoctest
+julia> gate = sigma_x(1)
+Gate Object: Snowflake.SigmaX
+Connected_qubits	: [1]
+Operator:
+(2,2)-element Snowflake.AntiDiagonalOperator:
+Underlying data type: ComplexF64:
+    .    1.0 + 0.0im
+    1.0 + 0.0im    .
+
+
+julia> move_gate(gate, [2])
+Gate Object: Snowflake.SigmaX
+Connected_qubits	: [2]
+Operator:
+(2,2)-element Snowflake.AntiDiagonalOperator:
+Underlying data type: ComplexF64:
+    .    1.0 + 0.0im
+    1.0 + 0.0im    .
+
+
+```
+"""
+function move_gate(gate::AbstractGate,
+    new_connected_qubits::AbstractVector{T})::AbstractGate where T<:Integer
+
+    return typeof(gate)(new_connected_qubits[1])
```

Comment:
This default is not a good default and could introduce a lot of very subtle bugs. If a default is provided it should work. It does not have to work optimally, but it should work. There are a lot of cases this will crash if it not specialize (all parameterized gates, all multi-qubit gates). There are also cases where this will produce the wrong output. For instance, if you create a gate with a variable number of qubits, such as a QFT gate, and forget to specialize this method then it won't crash and simply give a wrong answer. Bugs like that will be hard to track down.

It is also a lot of work to specialize this function for every gate that is not an unparameterized single qubit gate. Every-time someone creates a gate then they'll have to add this method as well. We can do this only once by creating a `MovedGate` this gates dispatches to the underlying gate for everything except for `get_connected_qubits`.

Reply:
Creating a `MovedGate` does remove the issue of having to write a `get_connected_qubits` function for each gate. However, a drawback of using `MovedGate` is that checking the gate type using `isa` returns `MovedGate` instead of the original gate type. This will cause problems in the transpiler. As a solution, we could create a function called `is_gate_type` which makes a type comparison with the original gate and not `MovedGate` directly.

Do you have another solution for this problem?

Reply:
I'm happy with having an `is_gate_type` function. It will also be useful if we are to add multiple gates with different names.

Reply:
The functions `is_gate_type` and `get_gate_type` have been added. For `AbstractGate` objects, they should be used instead of `isa` and `typeof`.